README.md
# UI Manager
*A UI Manager for PyQt & PySide*

**Copyright(c) 2021 Hkaar**

## Overview
UI Manger aims to help you in managing your ui more easily, and faster with less
code inside your program, with an easy to use interface so it's easy to use and learn
for use in your program.

## License
UI Manager is distributed under the MIT License. For more details please see
LICENSE.txt.

**Permission is hereby granted, free of charge, to any person obtaining a copy of 
this software and associated documentation files (the “Software”), to deal in 
the Software without restriction, including without limitation the rights to 
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies 
of the Software, and to permit persons to whom the Software is furnished to do 
so, subject to the following conditions:**

**The above copyright notice and this permission notice shall be included in all 
copies or substantial portions of the Software.**

**THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
SOFTWARE.**

## Installation
To install UI Manager on your device type these commands:
```
pip install ui-manager
```
or 
```
python setup.py build
python setup.py install
```

## How to use
UI Manager has a few classes to work with, and some functions. Here are some
examples on how to use them:
- Theme Manager:
```
import ui_manager

theme_manager = ui_manager.Theme_Manager()

theme_manager.import_theme('dark.qss', None, 'ascii') # import your theme.
theme_manager.set_theme('dark') # set your preffered theme.

theme_manager.set_widget_theme(widget1) # set the theme to your widget.
# or
theme_manager.add_widget(widget1) # add your widget to theme manager.
theme_manager.load_theme() # load the theme to your widget.
```
or
```
import ui_manager

theme_manager = ui_manager.Theme_Manager()

theme_manager.import_theme('dark.py', 'DarkTheme', None) # import your theme, from a class you select in the file.
theme_manager.set_theme('dark') # set your preffered theme.

theme_manager.set_widget_theme(widget1, 'default') # set the selected theme to your widget, with selecting a theme object from the theme.
# or
theme_manager.add_widget(widget1) # add your widget to theme manager.
theme_manager.load_theme() # load the theme to your widget.
```

## Dependecies
PyQt5 or PySide2


CHANGELOG.txt
# UI Manager CHANGELOG
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [0.2.0] - 2021-11-21
### Added
- Added support for .qss files for use in themes.
- Added support for importing files in folders (NOTE: this feature is only available for .qss files).
- Added support for importing more than one theme class in a .py file.
- Added error messages in some sections, for better debugging.

### Changed
- Improved support for importing .py files as themes.
- Moved default light and dark themes from classes to qss files.

### Removed
- Removed default light and dark theme classes.


old/__init__2.py
# PySide2 & PyQt5 UI Manager
# Made by Hkaar

import os

try:
    from PySide2.QtCore import QFile, QIODevice
except:
    from PyQt5.QtCore import QFile, QIODevice

class Theme_Manager:
    def __init__(self, theme='light'):
        self.stored_themes = {}
        self.theme = None

        self.stored_widgets = {}

        self.import_theme('/'.join(os.path.dirname(os.path.abspath(__file__)), 
            'resources/themes'), None)

    def set_theme(self, theme='light'):
        if isinstance(theme, str) and self.stored_themes.get(theme):
            self.theme = self.stored_themes[theme]
        else:
            raise KeyError((f"There is no such theme: {theme}, due "
                "to not being a string or theme does not exist!"))

    def import_theme(self, directory, class_name=None, encoding="ascii"):
        def append_theme(file_dir, class_name, encoding):
            file_name = os.path.basename(file_dir)

            if file_name.endswith('.py'):
                theme_f = __import__(file_name.replace('.py', ''))

                if class_name != None and hasattr(theme_f, class_name):
                    theme = getattr(theme_f, class_name)
                else:
                    raise TypeError((f"There is no such class: {class_name} "
                        f"inside {file_name} file!"))

                if hasattr(theme, 'name') and hasattr(theme, 'stylesheets'):
                    if theme.name in self.stored_themes:
                        self.stored_themes[theme.name][0].stylesheets.update(theme.stylesheets)
                    else:
                        self.stored_themes[theme.name] = (theme, 'py')
                else:
                    raise KeyError((f"Theme: {class_name} does not have 'name' & "
                        "'stylesheets' keys!"))

            elif file_name.endswith('.qss'):
                sty_f = QFile(file_dir)

                if class_name != None:
                    self.stored_themes[class_name] = (sty_f, 'qss', encoding)
                else:
                    self.stored_themes[file_name.replace('.qss', '')] = (sty_f, 'qss', encoding)

            else:
                raise ImportError((f"Failed to import: {file_name} due to invalid "
                    "format or file does not exist!"))

        def import_file(file_dir, class_name, encoding):        
            if isinstance(class_name, (list, tuple)) and file_dir.endswith('.py'):
                for cls_name in class_name:
                    append_theme(file_dir, cls_name, encoding)

            elif not isinstance(class_name, (list, tuple)):
                append_theme(file_dir, class_name, encoding)

            else:
                raise TypeError(f"Cannot use: {class_name} is considered an invalid type!")

        if os.path.isfile(directory) or directory.endswith('.qss'):
            import_file(directory, class_name, encoding)

        elif os.path.isdir(directory):
            files = ['/'.join((directory, f)) for f in os.listdir(directory) if os.path.isfile(
            '/'.join((directory, f))) and f.endswith('.qss')]

            for f in files:
                f_index = files.index(f)

                if isinstance(class_name, (list, tuple)):
                    if len(class_name)-1 >= f_index:
                        import_file(f, class_name[f_index], encoding)
                    else:
                        import_file(f, None, encoding)
                else:
                    import_file(f, class_name, encoding)

        else:
            raise FileNotFoundError(f"There is no such file or directory: {directory}!")

    def add_widget(self, widget, theme='default'):
        def gen_id():
            for widget_id in range(0, 9999):
                if not self.stored_widgets.get(widget_id):
                    return widget_id
            return None

        if isinstance(theme, str):
            if isinstance(widget, (list, tuple, set)):
                for w in widget:
                    self.stored_widgets[gen_id()] = (w, theme)
            else:
                self.stored_widgets[gen_id()] = (widget, theme)
        else:
            raise TypeError(f"Cannot use: {theme} is considered an invalid type!")

    def set_widget_theme(self, widget, theme='default'):
        def apply_theme(widget, theme):
            get_name = lambda obj: obj.__class__.__name__

            if self.theme[1] == 'py':
                if theme == 'palette' and hasattr(self.theme[0], 'palette'):
                    widget.setPalette(self.theme[0].palette)
                else:
                    widget.setStyleSheet(self.theme[0].stylesheets['default'])

                    if self.theme[0].stylesheets.get(theme):
                        widget.setStyleSheet(self.theme[0].stylesheets[theme])

                if get_name(widget) != 'QApplication':
                    widget.update()

            elif self.theme[1] == 'qss':
                sty_f = self.theme[0]

                sty_f.open(QIODevice.ReadOnly)
                widget.setStyleSheet(((sty_f.readAll()).data()).decode(self.theme[2]))

                if get_name(widget) != 'QApplication':
                    widget.update()

            else:
                raise KeyError((f"Current theme: {theme} has an invalid file type: "
                    f"{self.theme[1]}!"))

        if isinstance(theme, str):
            if isinstance(widget, (list, tuple, set)):
                for w in widget:
                    apply_theme(w, theme)
            else:
                apply_theme(widget, theme)
        else:
            raise TypeError(f"Cannot use: {theme} is considered an invalid type!")

    def load_theme(self):
        for w_id in self.stored_widgets:
            self.set_widget_theme(
                self.stored_widgets[w_id][0],
                self.stored_widgets[w_id][1]
                )

if __name__ == "__main__":
    th = Theme_Manager()
    th.import_theme('/home/hkaar/Desktop/test.qss', 'test')
    print(th.stored_themes)


old/__init__3_os.py
# PySide2 & PyQt5 UI Manager

__author__ = "Hkaar"
__version__ = 0.2

import os

try:
    from PySide2.QtCore import QFile, QIODevice
except ImportError:
    from PyQt5.QtCore import QFile, QIODevice

class Theme_Manager:
    def __init__(self, theme='light'):
        self.stored_themes = {}
        self.theme = None

        self.stored_widgets = {}

        self.import_theme('/'.join((os.path.dirname(os.path.abspath(__file__)
            ), 'resources/themes'),))
        self.set_theme(theme)

    def set_theme(self, theme='light'):
        if isinstance(theme, str) and self.stored_themes.get(theme):
            self.theme = self.stored_themes[theme]
        else:
            raise KeyError((f"There is no such theme: {theme}, due "
                "to not being a string or theme does not exist!"))

    def import_theme(self, directory, class_name=None, encoding='ascii'):
        def append_theme(file_dir, class_name, encoding):
            f_name = os.path.basename(file_dir)

            if f_name.endswith('.py'):
                theme_f = __import__(f_name.replace('.py', ''))

                if class_name != None and hasattr(theme_f, class_name):
                    theme = getattr(theme_f, class_name)
                else:
                    raise TypeError((f"There is no such class: {class_name} "
                        f"inside {f_name} file!"))

                if hasattr(theme, 'name') and hasattr(theme, 'stylesheets'):
                    if theme.name in self.stored_themes:
                        self.stored_themes[theme.name][0].stylesheets.update(theme.stylesheets)
                    else:
                        self.stored_themes[theme.name] = (theme, 'py')
                else:
                    raise KeyError((f"Theme: {class_name} does not have 'name' & "
                        "'stylesheets' keys!"))

            elif f_name.endswith('.qss'):
                sty_f = QFile(file_dir)

                if class_name != None:
                    self.stored_themes[class_name] = (sty_f, 'qss', encoding)
                else:
                    self.stored_themes[f_name.replace('.qss', '')] = (sty_f, 'qss', encoding)

            else:
                raise ImportError((f"Failed to import: {f_name} due to invalid "
                    "file format!"))

        def import_file(file_dir, class_name, encoding):
            if isinstance(class_name, (list, tuple)) and file_dir.endswith('.py'):
                for cls_name in class_name:
                    append_theme(file_dir, cls_name, encoding)

            elif not isinstance(class_name, (list, tuple)):
                append_theme(file_dir, class_name, encoding)

            else:
                raise TypeError(f"Cannot use: {class_name} is considered an invalid type!")

        if os.path.isfile(directory) or directory.endswith('.qss'):
            import_file(directory, class_name, encoding)

        elif os.path.isdir(directory):
            files = ['/'.join((directory, f)) for f in os.listdir(directory) if os.path.isfile(
            '/'.join((directory, f))) and f.endswith('.qss')]

            for f in files:
                f_index = files.index(f)

                if isinstance(class_name, (list, tuple)):
                    if len(class_name)-1 >= f_index:
                        import_file(f, class_name[f_index], encoding)
                    else:
                        import_file(f, None, encoding)
                else:
                    import_file(f, class_name, encoding)

        else:
            raise FileNotFoundError(f"There is no such file or directory: {directory}!")

    def add_widget(self, widget, theme='default'):
        def gen_id():
            for widget_id in range(0, 9999):
                if not self.stored_widgets.get(widget_id):
                    return widget_id
            return None

        if isinstance(theme, str):
            if isinstance(widget, (list, tuple, set)):
                for w in widget:
                    self.stored_widgets[gen_id()] = (w, theme)
            else:
                self.stored_widgets[gen_id()] = (widget, theme)
        else:
            raise TypeError(f"Cannot use: {theme} is considered an invalid type!")

    def set_widget_theme(self, widget, theme='default'):
        def apply_theme(widget, theme):
            get_name = lambda obj: obj.__class__.__name__

            if self.theme[1] == 'py':
                if theme == 'palette' and hasattr(self.theme[0], 'palette'):
                    widget.setPalette(self.theme[0].palette)
                else:
                    widget.setStyleSheet(self.theme[0].stylesheets['default'])

                    if self.theme[0].stylesheets.get(theme):
                        widget.setStyleSheet(self.theme[0].stylesheets[theme])

                if get_name(widget) != 'QApplication':
                    widget.update()

            elif self.theme[1] == 'qss':
                sty_f = self.theme[0]

                sty_f.open(QIODevice.ReadOnly)
                widget.setStyleSheet(((sty_f.readAll()).data()).decode(self.theme[2]))
                sty_f.close()

                if get_name(widget) != 'QApplication':
                    widget.update()

            else:
                raise KeyError((f"Current theme: {theme} has an invalid file type: "
                    f"{self.theme[1]}!"))

        if isinstance(theme, str):
            if isinstance(widget, (list, tuple, set)):
                for w in widget:
                    apply_theme(w, theme)
            else:
                apply_theme(widget, theme)
        else:
            raise TypeError(f"Cannot use: {theme} is considered an invalid type!")

    def load_theme(self):
        for w_id in self.stored_widgets:
            self.set_widget_theme(
                self.stored_widgets[w_id][0],
                self.stored_widgets[w_id][1]
                )


ui_manager/__init__.py
# PySide2 & PyQt5 UI Manager

__copyright__ = "Copyright (c) 2021 Hkaar"

__author__ = "Hkaar"
__version__ = 0.2

__license__ = "The MIT License (MIT)"

'''
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
associated documentation files (the “Software”), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, publish, distribute, 
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is 
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial 
portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT 
NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES 
OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
'''

from os import listdir, path

try:
    from PySide2.QtCore import QFile, QIODevice
except ImportError:
    from PyQt5.QtCore import QFile, QIODevice

class Theme_Manager:
    def __init__(self, theme='light'):
        self.stored_themes = {}
        self.theme = None

        self.stored_widgets = {}

        self.import_theme('/'.join((path.dirname(path.abspath(__file__)), 
            'resources/themes')),)
        self.set_theme(theme)

    def set_theme(self, theme='light'):
        if isinstance(theme, str) and self.stored_themes.get(theme):
            self.theme = self.stored_themes[theme]
        else:
            raise KeyError((f"There is no such theme: {theme}, due "
                "to not being a string or theme does not exist!"))

    def import_theme(self, directory, class_name=None, encoding='ascii'):
        def append_theme(file_dir, class_name, encoding):
            f_name = path.basename(file_dir)

            if f_name.endswith('.py'):
                theme_f = __import__(f_name.replace('.py', ''))

                if class_name != None and hasattr(theme_f, class_name):
                    theme = getattr(theme_f, class_name)
                else:
                    raise TypeError((f"There is no such class: {class_name} "
                        f"inside {f_name} file!"))

                if hasattr(theme, 'name') and hasattr(theme, 'stylesheets'):
                    if theme.name in self.stored_themes:
                        self.stored_themes[theme.name][0].stylesheets.update(theme.stylesheets)
                    else:
                        self.stored_themes[theme.name] = (theme, 'py')
                else:
                    raise KeyError((f"Theme: {class_name} does not have 'name' & "
                        "'stylesheets' keys!"))

            elif f_name.endswith('.qss'):
                sty_f = QFile(file_dir)

                if class_name != None:
                    self.stored_themes[class_name] = (sty_f, 'qss', encoding)
                else:
                    self.stored_themes[f_name.replace('.qss', '')] = (sty_f, 'qss', encoding)

            else:
                raise ImportError((f"Failed to import: {f_name} due to invalid "
                    "file format!"))

        def import_file(file_dir, class_name, encoding):
            if isinstance(class_name, (list, tuple)) and file_dir.endswith('.py'):
                for cls_name in class_name:
                    append_theme(file_dir, cls_name, encoding)

            elif not isinstance(class_name, (list, tuple)):
                append_theme(file_dir, class_name, encoding)

            else:
                raise TypeError(f"Cannot use: {class_name} is considered an invalid type!")

        if path.isfile(directory) or directory.endswith('.qss'):
            import_file(directory, class_name, encoding)

        elif path.isdir(directory):
            files = ['/'.join((directory, f)) for f in listdir(directory) if path.isfile(
            '/'.join((directory, f))) and f.endswith('.qss')]

            for f in files:
                f_index = files.index(f)

                if isinstance(class_name, (list, tuple)):
                    if len(class_name)-1 >= f_index:
                        import_file(f, class_name[f_index], encoding)
                    else:
                        import_file(f, None, encoding)
                else:
                    import_file(f, class_name, encoding)

        else:
            raise FileNotFoundError(f"There is no such file or directory: {directory}!")

    def add_widget(self, widget, theme='default'):
        def gen_id():
            for widget_id in range(0, 9999):
                if not self.stored_widgets.get(widget_id):
                    return widget_id
            return None

        if isinstance(theme, str):
            if isinstance(widget, (list, tuple, set)):
                for w in widget:
                    self.stored_widgets[gen_id()] = (w, theme)
            else:
                self.stored_widgets[gen_id()] = (widget, theme)
        else:
            raise TypeError(f"Cannot use: {theme} is considered an invalid type!")

    def set_widget_theme(self, widget, theme='default'):
        def apply_theme(widget, theme):
            get_name = lambda obj: obj.__class__.__name__

            if self.theme[1] == 'py':
                if theme == 'palette' and hasattr(self.theme[0], 'palette'):
                    widget.setPalette(self.theme[0].palette)
                else:
                    widget.setStyleSheet(self.theme[0].stylesheets['default'])

                    if self.theme[0].stylesheets.get(theme):
                        widget.setStyleSheet(self.theme[0].stylesheets[theme])

                if get_name(widget) != 'QApplication':
                    widget.update()

            elif self.theme[1] == 'qss':
                sty_f = self.theme[0]

                sty_f.open(QIODevice.ReadOnly)
                widget.setStyleSheet(((sty_f.readAll()).data()).decode(self.theme[2]))

                if get_name(widget) != 'QApplication':
                    widget.update()

                sty_f.close()

            else:
                raise KeyError((f"Current theme: {theme} has an invalid file type: "
                    f"{self.theme[1]}!"))

        if isinstance(theme, str):
            if isinstance(widget, (list, tuple, set)):
                for w in widget:
                    apply_theme(w, theme)
            else:
                apply_theme(widget, theme)
        else:
            raise TypeError(f"Cannot use: {theme} is considered an invalid type!")

    def load_theme(self):
        for w_id in self.stored_widgets:
            self.set_widget_theme(
                self.stored_widgets[w_id][0], 
                self.stored_widgets[w_id][1]
                )
